### PR TITLE
fix: separate cjs and mjs typings

### LINKED ISSUE #2010: Broken types using node16+ moduleResolution strategy with ESM
## 🐛 Bug Report

Using TypeScript with `moduleResolution` set to `node16` or `nodenext`, `module` set to any ESM module (such as `esnext`), and `"type": "module"` within the package.json, TypeScript is unable to correctly determine types for i18next imports due to it inferring the provided index.d.ts file as CJS when the underlying import is ESM, due to the single `index.d.ts` file being used for both ESM & CJS.

This is further explained on this page, https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseCJS.md

## To Reproduce

Reproduction steps:
* Using TypeScript with a tsconfig containing `"moduleResolution": "node16", "module": "esnext"`, and a package.json containing `"type": "module"`, the following code produces type errors

```js
import i18next from 'i18next';

// Property 'language' does not exist on type 'typeof import("/path/to/node_modules/i18next/index")'.
console.log(i18next.language);
```

## Expected behavior

There should be an index.d.ts file per entrypoint, so that it's correctly inferring CJS or ESM based on the types, and the types therefore work.

```js
import i18next from 'i18next';

// The `language` property is correctly seen as a string by TypeScript
console.log(i18next.language);
```

## Your Environment

- *runtime version*: Node 18.17.0
- *i18next version*: 23.4.1
- *os*: Mac
- *any other relevant information*: TypeScript version 5.1.6

### PR DESCRIPTION
**Pull request by JounQin**
close #2010

#### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [x] run tests `npm run test`
- [x] tests are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

#### Checklist (for documentation change)

- [ ] only relevant documentation part is changed (make a diff before you submit the PR)
- [ ] motivation/reason is provided
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by perrin4869** [comment]
This is related here: https://github.com/auth0/jwt-decode/pull/130/files
You can export a namespace by doing:

```
declare namespace jwtDecode {
  export { InvalidTokenError, JwtDecodeOptions, JwtHeader, JwtPayload }
}
export = jwtDecode
```

The PR was closed but it was the correct solution I believe

**Comment by JounQin** [comment]
> You can export a namespace by doing

@perrin4869 That's what I'm doing here.

**Comment by perrin4869** [comment]
right, sorry!

File: index.d.mts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import * as i18next from './index.js';
+
+import type { DefaultNamespace, Namespace } from './typescript/options.js';
+
+export type WithT<Ns extends Namespace = DefaultNamespace> = i18next.WithT<Ns>;
```
Comment (JounQin) [inline index.d.mts:5]: If you want, we can move these common typings into a separate `common.d.ts` file.

Comment (perrin4869) [reply index.d.mts:5]: Right, that's probably a good idea

Comment (JounQin) [reply index.d.mts:5]: @perrin4869 Interesting, the following would error:

```ts
// Internal Helpers
import type { WithT, i18n } from './typescript/common.js';

declare namespace i18next {
  export { WithT, i18n };
}

declare const i18next: i18n;

export = i18next;
```

image

[Playground](https://www.typescriptlang.org/play?target=99&jsx=1&module=1#code/JYWwDg9gTgLgBDAnmApnA3nA6sGALAFQBo5gBGADgDs4BfOAMyghDgHIA6AeiVQGcAxlGBgYXASxAQqHAFZ82AbgBQygCYoBAGwCGUNFR0gUfMDoFpy1FAA946ZXDi3IsDNlyESVmrRW1VDW09NAkqPngfWxgALlJKKhVlF2h4AF546ztFIA)

Any idea how to fix this to simplify?

File: package.json [hunk 2]
```diff
@@ -15,9 +15,14 @@
   "exports": {
     "./package.json": "./package.json",
     ".": {
-      "types": "./index.d.ts",
-      "import": "./dist/esm/i18next.js",
-      "require": "./dist/cjs/i18next.js"
+      "import": {
+        "types": "./index.d.mts",
```
Comment (me4502) [inline package.json:19]: If the two typing files are functionally identical, couldn't it just copy the root index.d.ts to both `./dist/esm/i18next.d.ts` and `./dist/cjs/i18next.d.ts` in the build script and direct to those here? It'd have minor duplication but the gzipping of the package would resolve most of that

Comment (JounQin) [reply package.json:19]: Nope, `export default` is incorrect for commonjs package.

File: index.d.mts (hunk 1 above)
Comment (perrin4869) [reply index.d.mts:5]: oh damn that's interesting... doesn't seem like that's going to work out...

**Comment by ubuntudroid** [comment]
@adrai @pedrodurek anything preventing this from being merged? My project is suffering from this type issue as well and I would love to get my hands on this fix. 😎

**Comment by adrai** [comment]
@JounQin is this PR ready to review?

**Comment by JounQin** [comment]
> @JounQin is this PR ready to review?

Yes, although I haven't rebased due to lack of reviews.

**Comment by adrai** [comment]
@pedrodurek is there a chance you could review this?

**Comment by ubuntudroid** [comment]
As @pedrodurek doesn't seem to be available at the moment, can someone else take over the final review?

**Comment by adrai** [comment]
@nicegamer7 are you able to review this?

**Comment by adrai** [comment]
@JounQin can you rebase this?
btw: why are you importing via `.js` extension and not via `.d.ts`? Sorry for my naive question.

**Comment by ubuntudroid** [comment]
Any updates here? 🙂

**Comment by nicegamer7** [comment]
Was waiting for this to be rebased. @adrai should I review this anyway without the rebase?

**Comment by adrai** [comment]
@JounQin ?

**Comment by JounQin** [comment]
Ah, sorry, I'm on a vocation these days, would be free maybe tomorrow.

**Comment by JounQin** [comment]
> @JounQin can you rebase this? btw: why are you importing via `.js` extension and not via `.d.ts`? Sorry for my naive question.

@adrai Just rebased.

`.js` works with `.js`/`.ts`/`.d.ts` files for all resolutions, `.d.ts` could stop working for maybe node16+ resolutions.

**Comment by adrai** [comment]
> `.js` works with `.js`/`.ts`/`.d.ts` files for all resolutions, `.d.ts` could stop working for maybe node16+ resolutions.

So importing .js works even though the imported file has a .d.ts extension? Can you share some information/link about why it maybe will stop working in future?

**Comment by adrai** [comment]
https://github.com/i18next/i18next/issues/2048

**Comment by adrai** [comment]
I solved it without a namespace, because of issues with interpolation variable validation etc....
please try with v23.7.0

**Comment by adrai** [comment]
Seems mts is generating some problems here: https://github.com/i18next/react-i18next/commit/a74282e218d408715864d04f7057b904564bc488#commitcomment-132242258
Would be awesome if anyone could help me to update react-i18next to also separate cjs and mjs typings => https://github.com/i18next/react-i18next/tree/mts

**Comment by JounQin** [comment]
Don't use TSLint which does not `.mts` at all.

**Comment by adrai** [comment]
> Don't use TSLint which does not `.mts` at all.

I don't think it's about tslint

**Comment by adrai** [comment]
ok, this seems to fix it: https://github.com/i18next/i18next/commit/85fd79b390c8d7a01b5496a257520bd78e9da2eb

**Pull request closed without merging**